File: app/src/main/java/com/taiwanlife/teamwalk/service/TWFirebaseMessagingService.java
/**
 * Copyright (c) 1989-2020 INVENTEC BESTA Co., Ltd.
 */
package com.taiwanlife.teamwalk.service;

import android.app.PendingIntent;
import android.content.Intent;

import androidx.annotation.NonNull;
import androidx.core.app.NotificationCompat;
import androidx.core.app.NotificationManagerCompat;

import com.google.firebase.messaging.FirebaseMessagingService;
import com.google.firebase.messaging.RemoteMessage;
import com.taiwanlife.teamwalk.MainActivity;
import com.taiwanlife.teamwalk.R;

/**
 * @author Vincent.Chen
 * @version 1
 *
 * @date 2020/11/2
 */
public class TWFirebaseMessagingService extends FirebaseMessagingService {

    private static final String DEFAULT_CHANNEL = "default";

    /**
     *
     * @param remoteMessage
     */
    @Override
    public void onMessageReceived(@NonNull RemoteMessage remoteMessage) {
        super.onMessageReceived(remoteMessage);

        if (remoteMessage.getNotification() != null) {
            RemoteMessage.Notification notification = remoteMessage.getNotification();
            Intent intent = new Intent(this, MainActivity.class);
            intent.putExtra("url", remoteMessage.getData().get("url"));
            intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_CLEAR_TASK);
            PendingIntent pendingIntent;

            if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.S) {
                pendingIntent = PendingIntent.getActivity(this, (int)remoteMessage.getSentTime(), intent, PendingIntent.FLAG_IMMUTABLE);
            } else {
                pendingIntent = PendingIntent.getActivity(this, (int)remoteMessage.getSentTime(), intent, PendingIntent.FLAG_UPDATE_CURRENT);
            }


            NotificationCompat.Builder builder = new NotificationCompat.Builder(this, getString(R.string.noti_channel_id))
                    .setSmallIcon(R.drawable.ic_firebase)
                    .setContentTitle(notification.getTitle())
                    .setContentText(notification.getBody())
                    .setPriority(NotificationCompat.PRIORITY_HIGH)
                    .setContentIntent(pendingIntent)
                    .setAutoCancel(true)
                    .setVibrate(new long[0]);

            NotificationManagerCompat notificationManager = NotificationManagerCompat.from(this);
            notificationManager.notify((int)remoteMessage.getSentTime(), builder.build());
        }
    }
}

File: app/src/main/java/com/taiwanlife/teamwalk/service/TWFirebaseMessagingService.kt
package com.taiwanlife.teamwalk.service

import android.app.PendingIntent
import android.content.Intent
import android.os.Build
import androidx.core.net.toUri
import com.google.firebase.messaging.FirebaseMessagingService
import com.google.firebase.messaging.RemoteMessage
import com.taiwanlife.teamwalk.Config
import com.taiwanlife.teamwalk.Config.NOTIFICATION_KEY_BADGE
import com.taiwanlife.teamwalk.Config.NOTIFICATION_KEY_MSG
import com.taiwanlife.teamwalk.Config.NOTIFICATION_KEY_TITLE
import com.taiwanlife.teamwalk.Config.NOTIFICATION_KEY_TYPE
import com.taiwanlife.teamwalk.Config.NOTIFICATION_KEY_URL
import com.taiwanlife.teamwalk.ui.main.MainActivity
import com.taiwanlife.teamwalk.utils.MyNotificationManager
import com.taiwanlife.teamwalk.utils.SecuredPreferenceStoreManager
import com.taiwanlife.teamwalk.utils.Utils.stringToNotificationType

class TWFirebaseMessagingService : FirebaseMessagingService() {
    /**
     *
     * @param remoteMessage
     */
    override fun onMessageReceived(remoteMessage: RemoteMessage) {
        super.onMessageReceived(remoteMessage)

        val intent = Intent(this, MainActivity::class.java)
        val msg = remoteMessage.getData()[NOTIFICATION_KEY_MSG]
        val url = remoteMessage.getData()[NOTIFICATION_KEY_URL]
        val type = remoteMessage.getData()[NOTIFICATION_KEY_TYPE]
        val badge = remoteMessage.getData()[NOTIFICATION_KEY_BADGE]
        val title = remoteMessage.getData()[NOTIFICATION_KEY_TITLE]

        val myNotificationManager = MyNotificationManager(this)

        if (!badge.isNullOrEmpty() && badge.toInt() > 0) {
            // 更新未讀數量
            val pendingIntent = MyNotificationManager.getPendingIntentForBadge(
                this,
                remoteMessage.sentTime.toInt(),
                badge.toInt(),
                title,
                msg,
                type,
                url
            )
            myNotificationManager.updateBadge(pendingIntent, badge.toInt())
            // 如果要使用後端帶給我們的Title和Msg
//            myNotificationManager.updateBadge(pendingIntent, badge.toInt(), title, msg)
        }

        val notificationType = stringToNotificationType(type)
        val flags = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
            PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE
        } else {
            PendingIntent.FLAG_UPDATE_CURRENT
        }

        val pendingIntent = if (notificationType == Config.NotificationType.URL /* "L" */ && !url.isNullOrEmpty()) {
            // 收到需要外開瀏覽器的URL的網址 準備交由外面瀏覽器開啟
            val browserIntent = Intent(Intent.ACTION_VIEW, url.toUri())
            // 先檢查是否有 Chrome 並將Package指定由Chrome開啟(Explicit Intent)
            val chromePackage = "com.android.chrome"
            browserIntent.setPackage(chromePackage)

            val finalIntent = if (isPackageInstalled(chromePackage)) {
                // 如果有 Chrome，直接使用
                browserIntent
            } else {
                // 如果沒有 Chrome，清除 package 設定並建立選擇器(Intent Chooser)
                browserIntent.setPackage(null)
                Intent.createChooser(browserIntent, "請選擇瀏覽器")
            }
            PendingIntent.getActivity(
                this,
                remoteMessage.getSentTime().toInt(),
                finalIntent,
                flags
            )
        } else {
            intent.putExtra(NOTIFICATION_KEY_URL, url)
            intent.putExtra(NOTIFICATION_KEY_TYPE, type)
            intent.putExtra(NOTIFICATION_KEY_TITLE, title)
            intent.putExtra(NOTIFICATION_KEY_MSG, msg)
            intent.flags = Intent.FLAG_ACTIVITY_SINGLE_TOP or Intent.FLAG_ACTIVITY_CLEAR_TOP

            PendingIntent.getActivity(
                this,
                remoteMessage.getSentTime().toInt(),
                intent,
                flags
            )
        }
        val notificationId = remoteMessage.sentTime.toInt()
        myNotificationManager.sendFcmNotification(notificationId, title, msg, pendingIntent)
    }

    private fun isPackageInstalled(packageName: String): Boolean {
        return try {
            packageManager.getPackageInfo(packageName, 0)
            true
        } catch (e: Exception) {
            false
        }
    }

    override fun onNewToken(token: String) {
        super.onNewToken(token)

        SecuredPreferenceStoreManager.simpleEditAndApply(
            Config.SP_FCM_IDENTIFIER,
            token
        )
    }
}